x/cdp/keeper/auctions.go

```diff
@@ -0,0 +1,189 @@
+	return partialDeposit{
+		Depositor: depositor,
+		Amount:    amount,
+		DebtShare: ds,
+	}
+}
+
+type partialDeposits []partialDeposit
+
+func (pd partialDeposits) SumCollateral() (sum sdk.Int) {
+	sum = sdk.ZeroInt()
+	for _, d := range pd {
+		sum = sum.Add(d.Amount[0].Amount)
+	}
+	return
+}
+
+func (pd partialDeposits) SumDebt() (sum sdk.Int) {
+	sum = sdk.ZeroInt()
+	for _, d := range pd {
+		sum = sum.Add(d.DebtShare)
+	}
+	return
+}
+
+// AuctionCollateral creates auctions from the input deposits which attempt to raise the corresponding amount of debt
+func (k Keeper) AuctionCollateral(ctx sdk.Context, deposits types.Deposits, debt sdk.Int, bidDenom string) {
+	auctionSize := k.getAuctionSize(ctx, deposits[0].Amount[0].Denom)
+	partialAuctionDeposits := partialDeposits{}
+	totalCollateral := deposits.SumCollateral()
+	for deposits.SumCollateral().GTE(auctionSize) {
+		for i, dep := range deposits {
+			// create auctions from individual deposits that are larger than the auction size
+			k.CreateAuctionsFromDeposit(ctx, &dep, &debt, &totalCollateral, auctionSize, bidDenom)
+			if !dep.Amount[0].Amount.IsZero() {
+				debtCoveredByDeposit := (dep.Amount[0].Amount.Quo(totalCollateral)).Mul(debt)
+				if (partialAuctionDeposits.SumCollateral().Add(dep.Amount[0].Amount)).LT(auctionSize) {
+					pd := newPartialDeposit(dep.Depositor, dep.Amount, debtCoveredByDeposit)
+					partialAuctionDeposits = append(partialAuctionDeposits, pd)
+					dep.Amount = sdk.NewCoins(sdk.NewCoin(dep.Amount[0].Denom, sdk.ZeroInt()))
+				} else {
+					partialCollateral := sdk.NewCoins(sdk.NewCoin(dep.Amount[0].Denom, auctionSize.Sub(partialAuctionDeposits.SumCollateral())))
+					partialDebt := (partialCollateral[0].Amount.Quo(dep.Amount[0].Amount)).Mul(debtCoveredByDeposit)
+
+					partialDep := newPartialDeposit(dep.Depositor, partialCollateral, partialDebt)
+					partialAuctionDeposits = append(partialAuctionDeposits, partialDep)
+					k.CreateAuctionFromPartialDeposits(ctx, partialAuctionDeposits, &debt, &totalCollateral, auctionSize, bidDenom)
+					partialAuctionDeposits = partialDeposits{}
+					dep.Amount = sdk.NewCoins(sdk.NewCoin(dep.Amount[0].Denom, dep.Amount[0].Amount.Sub(partialCollateral[0].Amount)))
+				}
+			}
+			if dep.Amount.IsZero() {
+				// remove the deposit from the slice if it is empty
+				deposits = append(deposits[:i], deposits[i+1:]...)
+				i--
+			} else {
+				deposits[i] = dep
+			}
+		}
+	}
+	if partialAuctionDeposits.SumCollateral().GT(sdk.ZeroInt()) {
+		k.CreateAuctionFromPartialDeposits(ctx, partialAuctionDeposits, &debt, &totalCollateral, partialAuctionDeposits.SumCollateral(), bidDenom)
+	}
+}
+
+// CreateAuctionsFromDeposit creates auctions from the input deposit until there is less than auctionSize left on the deposit
+func (k Keeper) CreateAuctionsFromDeposit(ctx sdk.Context, dep *types.Deposit, debt *sdk.Int, totalCollateral *sdk.Int, auctionSize sdk.Int, principalDenom string) {
```

Comment:
This method would be simpler if it used local variables to keep track of the data instead of updating the state via ref in each iteration of the loop. Return local variables and use them to update the state back in the primary function.